File: lib/moof/normalizer.rb
require_relative "ast"

module Moof
  class Normalizer
    def call(program)
      AST::Program.new(expressions: program.expressions.map { |e| normalize(e) })
    end

    private

    def normalize(node)
      case node
      when AST::MessageSend
        receiver = normalize(node.receiver)
        args = node.arguments.map { |a| normalize(a) }
        AST::Call.new(
          callee: AST::Identifier.new(name: "__send", line: node.line, column: node.column),
          arguments: [receiver, AST::StringLiteral.new(value: node.selector, line: node.line, column: node.column), *args],
          line: node.line, column: node.column
        )
      when AST::Call
        AST::Call.new(callee: normalize(node.callee),
          arguments: node.arguments.map { |a| normalize(a) },
          line: node.line, column: node.column)
      when AST::Define
        AST::Define.new(name: node.name, value: normalize(node.value), line: node.line, column: node.column)
      when AST::DefineFunction
        AST::DefineFunction.new(name: node.name, params: node.params, rest_param: node.rest_param,
          body: normalize(node.body), line: node.line, column: node.column)
      when AST::Lambda
        AST::Lambda.new(params: node.params, rest_param: node.rest_param,
          body: normalize(node.body), line: node.line, column: node.column)
      when AST::If
        AST::If.new(condition: normalize(node.condition),
          then_branch: normalize(node.then_branch),
          else_branch: node.else_branch ? normalize(node.else_branch) : nil,
          line: node.line, column: node.column)
      when AST::Let
        bindings = node.bindings.map { |name, val| [name, normalize(val)] }
        AST::Let.new(bindings: bindings, body: normalize(node.body), line: node.line, column: node.column)
      when AST::Do
        AST::Do.new(expressions: node.expressions.map { |e| normalize(e) }, line: node.line, column: node.column)
      when AST::SetBang
        AST::SetBang.new(name: node.name, value: normalize(node.value), line: node.line, column: node.column)
      when AST::Quote
        node
      when AST::TryCatch
        AST::TryCatch.new(body: normalize(node.body), error_name: node.error_name,
          catch_body: normalize(node.catch_body), line: node.line, column: node.column)
      when AST::Cond
        clauses = node.clauses.map do |test_node, body_node|
          if test_node == :else || (test_node.is_a?(AST::Identifier) && test_node.name == "else")
            [:else, normalize(body_node)]
          else
            [normalize(test_node), normalize(body_node)]
          end
        end
        AST::Cond.new(clauses: clauses, line: node.line, column: node.column)
      when AST::And
        AST::And.new(left: normalize(node.left), right: normalize(node.right), line: node.line, column: node.column)
      when AST::Or
        AST::Or.new(left: normalize(node.left), right: normalize(node.right), line: node.line, column: node.column)
      when AST::MapLiteral
        pairs = node.pairs.map { |key, val| [key, normalize(val)] }
        AST::MapLiteral.new(pairs: pairs, line: node.line, column: node.column)
      when AST::Program
        AST::Program.new(expressions: node.expressions.map { |e| normalize(e) })
      when AST::ClassDef
        methods = node.methods.map { |m|
          AST::MethodDef.new(selector: m.selector, params: m.params, body: normalize(m.body),
            line: m.line, column: m.column)
        }
        AST::ClassDef.new(name: node.name, superclass: node.superclass, fields: node.fields,
          methods: methods, traits: node.traits, line: node.line, column: node.column)
      when AST::TraitDef
        methods = node.methods.map { |m|
          AST::MethodDef.new(selector: m.selector, params: m.params, body: normalize(m.body),
            line: m.line, column: m.column)
        }
        AST::TraitDef.new(name: node.name, methods: methods, line: node.line, column: node.column)

      # ============ New node normalization ============

      # Pipeline: desugar (-> val step1 step2) into nested calls
      when AST::Pipeline
        desugar_pipeline(node)

      # StringInterp: desugar into nested concat calls
      when AST::StringInterp
        desugar_string_interp(node)

      # SelectorRef: desugar &name into lambda
      when AST::SelectorRef
        desugar_selector_ref(node)

      # KeywordArg: normalize the value
      when AST::KeywordArg
        AST::KeywordArg.new(keyword: node.keyword, value: normalize(node.value),
          line: node.line, column: node.column)

      # Match: recursively normalize children
      when AST::Match
        clauses = node.clauses.map do |clause|
          AST::MatchClause.new(
            pattern: normalize_pattern(clause.pattern),
            guard: clause.guard ? normalize(clause.guard) : nil,
            body: normalize(clause.body)
          )
        end
        AST::Match.new(expr: normalize(node.expr), clauses: clauses, line: node.line, column: node.column)

      # TypeDef: pass through (no desugaring needed)
      when AST::TypeDef
        node

      # ProtocolDef: pass through
      when AST::ProtocolDef
        node

      # DefMacro: recursively normalize body
      when AST::DefMacro
        AST::DefMacro.new(name: node.name, params: node.params,
          body: normalize(node.body), line: node.line, column: node.column)

      # Quasiquote: recursively normalize
      when AST::Quasiquote
        AST::Quasiquote.new(expression: normalize(node.expression), line: node.line, column: node.column)

      when AST::Unquote
        AST::Unquote.new(expression: normalize(node.expression), line: node.line, column: node.column)

      when AST::UnquoteSplice
        AST::UnquoteSplice.new(expression: normalize(node.expression), line: node.line, column: node.column)

      when AST::ModuleDef
        AST::ModuleDef.new(name: node.name, exports: node.exports,
          body: node.body.map { |e| normalize(e) }, line: node.line, column: node.column)

      when AST::UseModule, AST::Require
        node

      else
        node
      end
    end

    # Normalize pattern nodes (most are leaf nodes, but some contain sub-patterns)
    def normalize_pattern(pat)
      case pat
      when AST::MatchList
        elements = pat.elements.map { |e| normalize_pattern(e) }
        rest = pat.rest ? normalize_pattern(pat.rest) : nil
        AST::MatchList.new(elements: elements, rest: rest, line: pat.line, column: pat.column)
      when AST::MatchMap
        pairs = pat.pairs.map { |k, p| [k, normalize_pattern(p)] }
        AST::MatchMap.new(pairs: pairs, line: pat.line, column: pat.column)
      when AST::MatchConstructor
        bindings = pat.bindings.map { |b| normalize_pattern(b) }
        AST::MatchConstructor.new(class_name: pat.class_name, bindings: bindings,
          line: pat.line, column: pat.column)
      else
        pat
      end
    end

    # Desugar (-> val step1 step2) into nested calls
    # For each step:
    #   - MessageSend: inject previous as receiver
    #   - Call: inject previous as first argument
    #   - Identifier: wrap in Call(step, [previous])
    def desugar_pipeline(node)
      result = normalize(node.value)
      node.steps.each do |step|
        result = pipeline_inject(step, result, node.line, node.column)
      end
      result
    end

    def pipeline_inject(step, prev, ln, col)
      case step
      when AST::MessageSend
        # Inject prev as receiver, then normalize
        injected = AST::MessageSend.new(
          receiver: prev, selector: step.selector,
          arguments: step.arguments, line: step.line, column: step.column
        )
        normalize(injected)
      when AST::Call
        # Inject prev as first argument
        normalized_callee = normalize(step.callee)
        normalized_args = step.arguments.map { |a| normalize(a) }
        AST::Call.new(
          callee: normalized_callee,
          arguments: [prev, *normalized_args],
          line: step.line, column: step.column
        )
      when AST::Identifier
        # Wrap in Call(step, [prev])
        AST::Call.new(
          callee: step,
          arguments: [prev],
          line: step.line, column: step.column
        )
      else
        # For any other expression type, wrap as a call with prev as argument
        normalized_step = normalize(step)
        AST::Call.new(
          callee: normalized_step,
          arguments: [prev],
          line: ln, column: col
        )
      end
    end

    # Desugar $"hello \(name)" into nested concat calls via __send
    def desugar_string_interp(node)
      ln, col = node.line, node.column
      parts = node.segments.map do |seg|
        normalized = normalize(seg)
        if seg.is_a?(AST::StringLiteral)
          normalized
        else
          # Call to_s on expression: (__send expr "to_s")
          make_send(normalized, "to_s", [], ln, col)
        end
      end

      return AST::StringLiteral.new(value: "", line: ln, column: col) if parts.empty?
      return parts.first if parts.length == 1

      # Chain concat calls: (__send (__send part1 "concat:" part2) "concat:" part3)
      result = parts.first
      parts[1..].each do |part|
        result = make_send(result, "concat:", [part], ln, col)
      end
      result
    end

    # Desugar &name into lambda
    def desugar_selector_ref(node)
      ln, col = node.line, node.column
      obj_param = "__obj"
      obj_id = AST::Identifier.new(name: obj_param, line: ln, column: col)

      body = make_send(obj_id, node.selector, node.partial_args.map { |a| normalize(a) }, ln, col)
      AST::Lambda.new(params: [obj_param], body: body, line: ln, column: col)
    end

    # Helper: make a __send call
    def make_send(receiver, selector, args, ln, col)
      AST::Call.new(
        callee: AST::Identifier.new(name: "__send", line: ln, column: col),
        arguments: [receiver, AST::StringLiteral.new(value: selector, line: ln, column: col), *args],
        line: ln, column: col
      )
    end
  end
end

File: src/normalizer.rs
use crate::ast::*;

pub fn normalize(program: Program) -> Program {
    Program {
        expressions: program.expressions.into_iter().map(normalize_expr).collect(),
    }
}

fn normalize_expr(expr: Expr) -> Expr {
    match expr {
        // MessageSend -> Call(__send, [receiver, selector_str, ...args])
        Expr::MessageSend(receiver, selector, args, loc) => {
            let recv = normalize_expr(*receiver);
            let norm_args: Vec<Expr> = args.into_iter().map(normalize_expr).collect();
            let ln = loc.line.unwrap_or(0);
            let col = loc.column.unwrap_or(0);
            make_send(recv, &selector, norm_args, ln, col)
        }

        // Call: normalize callee and args recursively
        Expr::Call(callee, args, loc) => {
            let callee = normalize_expr(*callee);
            let args: Vec<Expr> = args.into_iter().map(normalize_expr).collect();
            Expr::Call(Box::new(callee), args, loc)
        }

        // Define
        Expr::Define(name, value, loc) => {
            Expr::Define(name, Box::new(normalize_expr(*value)), loc)
        }

        // Lambda
        Expr::Lambda(params, rest, body, loc) => {
            Expr::Lambda(params, rest, Box::new(normalize_expr(*body)), loc)
        }

        // If
        Expr::If(cond, then_br, else_br, loc) => {
            Expr::If(
                Box::new(normalize_expr(*cond)),
                Box::new(normalize_expr(*then_br)),
                else_br.map(|e| Box::new(normalize_expr(*e))),
                loc,
            )
        }

        // Let
        Expr::Let(bindings, body, loc) => {
            let bindings = bindings
                .into_iter()
                .map(|(name, val)| (name, normalize_expr(val)))
                .collect();
            Expr::Let(bindings, Box::new(normalize_expr(*body)), loc)
        }

        // Do
        Expr::Do(exprs, loc) => {
            Expr::Do(exprs.into_iter().map(normalize_expr).collect(), loc)
        }

        // SetBang
        Expr::SetBang(name, value, loc) => {
            Expr::SetBang(name, Box::new(normalize_expr(*value)), loc)
        }

        // Quote: do NOT recurse into contents
        e @ Expr::Quote(_, _) => e,

        // Quasiquote: pass through (do not normalize contents per Ruby prototype comment)
        Expr::Quasiquote(inner, loc) => {
            Expr::Quasiquote(Box::new(normalize_expr(*inner)), loc)
        }

        // Unquote
        Expr::Unquote(inner, loc) => {
            Expr::Unquote(Box::new(normalize_expr(*inner)), loc)
        }

        // UnquoteSplice
        Expr::UnquoteSplice(inner, loc) => {
            Expr::UnquoteSplice(Box::new(normalize_expr(*inner)), loc)
        }

        // TryCatch
        Expr::TryCatch(body, err_name, catch_body, loc) => {
            Expr::TryCatch(
                Box::new(normalize_expr(*body)),
                err_name,
                Box::new(normalize_expr(*catch_body)),
                loc,
            )
        }

        // MapLiteral
        Expr::MapLiteral(pairs, loc) => {
            let pairs = pairs
                .into_iter()
                .map(|(k, v)| (k, Box::new(normalize_expr(*v))))
                .collect();
            Expr::MapLiteral(pairs, loc)
        }

        // ClassDef
        Expr::ClassDef { name, superclass, fields, methods, traits, loc } => {
            let methods = methods
                .into_iter()
                .map(|m| MethodDef {
                    selector: m.selector,
                    params: m.params,
                    body: Box::new(normalize_expr(*m.body)),
                    loc: m.loc,
                })
                .collect();
            Expr::ClassDef { name, superclass, fields, methods, traits, loc }
        }

        // TraitDef
        Expr::TraitDef { name, methods, loc } => {
            let methods = methods
                .into_iter()
                .map(|m| MethodDef {
                    selector: m.selector,
                    params: m.params,
                    body: Box::new(normalize_expr(*m.body)),
                    loc: m.loc,
                })
                .collect();
            Expr::TraitDef { name, methods, loc }
        }

        // Pipeline: desugar into nested calls
        Expr::Pipeline(value, steps, loc) => {
            desugar_pipeline(*value, steps, &loc)
        }

        // StringInterp: desugar into chained concat calls
        Expr::StringInterp(segments, loc) => {
            desugar_string_interp(segments, &loc)
        }

        // SelectorRef: desugar into lambda
        Expr::SelectorRef(selector, partial_args, loc) => {
            desugar_selector_ref(&selector, partial_args, &loc)
        }

        // Match
        Expr::Match(expr, clauses, loc) => {
            let clauses = clauses
                .into_iter()
                .map(|c| MatchClause {
                    pattern: normalize_pattern(c.pattern),
                    guard: c.guard.map(|g| Box::new(normalize_expr(*g))),
                    body: Box::new(normalize_expr(*c.body)),
                })
                .collect();
            Expr::Match(Box::new(normalize_expr(*expr)), clauses, loc)
        }

        // TypeDef: pass through
        e @ Expr::TypeDef(_, _, _) => e,

        // ProtocolDef: pass through
        e @ Expr::ProtocolDef(_, _, _) => e,

        // DefMacro: normalize body
        Expr::DefMacro(name, params, body, loc) => {
            Expr::DefMacro(name, params, Box::new(normalize_expr(*body)), loc)
        }

        // ModuleDef
        Expr::ModuleDef(name, exports, body, loc) => {
            let body = body.into_iter().map(normalize_expr).collect();
            Expr::ModuleDef(name, exports, body, loc)
        }

        // UseModule, Require: pass through
        e @ Expr::UseModule(_, _, _, _) => e,
        e @ Expr::Require(_, _) => e,

        // Leaf nodes: pass through unchanged
        e @ (Expr::Integer(_, _)
        | Expr::Float(_, _)
        | Expr::Str(_, _)
        | Expr::Bool(_, _)
        | Expr::Nil(_)
        | Expr::Identifier(_, _)) => e,
    }
}

/// Normalize pattern nodes (most are leaf nodes, but some contain sub-patterns).
fn normalize_pattern(pat: Pattern) -> Pattern {
    match pat {
        Pattern::List(elements, rest) => {
            let elements = elements.into_iter().map(normalize_pattern).collect();
            let rest = rest.map(|r| Box::new(normalize_pattern(*r)));
            Pattern::List(elements, rest)
        }
        Pattern::Map(pairs) => {
            let pairs = pairs
                .into_iter()
                .map(|(k, p)| (k, normalize_pattern(p)))
                .collect();
            Pattern::Map(pairs)
        }
        Pattern::Constructor(name, bindings) => {
            let bindings = bindings.into_iter().map(normalize_pattern).collect();
            Pattern::Constructor(name, bindings)
        }
        // Wildcard, Bind, Literal: pass through
        other => other,
    }
}

// ── Desugaring helpers ───────────────────────────────────────────────

/// Desugar (-> val step1 step2) into nested calls.
fn desugar_pipeline(value: Expr, steps: Vec<Expr>, loc: &Loc) -> Expr {
    let mut result = normalize_expr(value);
    let ln = loc.line.unwrap_or(0);
    let col = loc.column.unwrap_or(0);
    for step in steps {
        result = pipeline_inject(step, result, ln, col);
    }
    result
}

fn pipeline_inject(step: Expr, prev: Expr, ln: usize, col: usize) -> Expr {
    match step {
        // MessageSend: inject prev as receiver, then normalize
        Expr::MessageSend(_, selector, args, step_loc) => {
            let injected = Expr::MessageSend(
                Box::new(prev),
                selector,
                args,
                step_loc,
            );
            normalize_expr(injected)
        }
        // Call: inject prev as first argument
        Expr::Call(callee, args, step_loc) => {
            let normalized_callee = normalize_expr(*callee);
            let mut normalized_args = vec![prev];
            normalized_args.extend(args.into_iter().map(normalize_expr));
            Expr::Call(Box::new(normalized_callee), normalized_args, step_loc)
        }
        // Identifier: wrap as Call(step, [prev])
        e @ Expr::Identifier(_, _) => {
            Expr::Call(
                Box::new(e),
                vec![prev],
                Loc::new(ln, col),
            )
        }
        // Any other expression: wrap as a call with prev as argument
        other => {
            let normalized_step = normalize_expr(other);
            Expr::Call(
                Box::new(normalized_step),
                vec![prev],
                Loc::new(ln, col),
            )
        }
    }
}

/// Desugar $"hello \(name)" into chained concat calls via __send.
fn desugar_string_interp(segments: Vec<Expr>, loc: &Loc) -> Expr {
    let ln = loc.line.unwrap_or(0);
    let col = loc.column.unwrap_or(0);

    let parts: Vec<Expr> = segments
        .into_iter()
        .map(|seg| {
            let is_str = matches!(&seg, Expr::Str(_, _));
            let normalized = normalize_expr(seg);
            if is_str {
                normalized
            } else {
                // Call to_s on expression: (__send expr "to_s")
                make_send(normalized, "to_s", vec![], ln, col)
            }
        })
        .collect();

    if parts.is_empty() {
        return Expr::Str(String::new(), Loc::new(ln, col));
    }
    if parts.len() == 1 {
        return parts.into_iter().next().unwrap();
    }

    // Chain concat calls: (__send (__send part1 "concat:" part2) "concat:" part3)
    let mut iter = parts.into_iter();
    let mut result = iter.next().unwrap();
    for part in iter {
        result = make_send(result, "concat:", vec![part], ln, col);
    }
    result
}

/// Desugar &name into lambda wrapping __send.
fn desugar_selector_ref(selector: &str, partial_args: Vec<Expr>, loc: &Loc) -> Expr {
    let ln = loc.line.unwrap_or(0);
    let col = loc.column.unwrap_or(0);
    let obj_param = "__obj".to_string();
    let obj_id = Expr::Identifier(obj_param.clone(), Loc::new(ln, col));

    let norm_args: Vec<Expr> = partial_args.into_iter().map(normalize_expr).collect();
    let body = make_send(obj_id, selector, norm_args, ln, col);

    Expr::Lambda(vec![obj_param], None, Box::new(body), Loc::new(ln, col))
}

/// Helper: construct a __send call node.
fn make_send(receiver: Expr, selector: &str, args: Vec<Expr>, ln: usize, col: usize) -> Expr {
    let mut all_args = vec![
        receiver,
        Expr::Str(selector.to_string(), Loc::new(ln, col)),
    ];
    all_args.extend(args);
    Expr::Call(
        Box::new(Expr::Identifier("__send".to_string(), Loc::new(ln, col))),
        all_args,
        Loc::new(ln, col),
    )
}
